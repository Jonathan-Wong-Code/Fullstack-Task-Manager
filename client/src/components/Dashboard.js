import React, { useEffect, useState } from "react";
import TaskList from "./TaskList";
import Pagination from "./Pagination";
import { useTaskState, useTaskDispatch } from "./../context/task-context";
import FilterBar from "./FilterBar";
import { fetchAllTasks } from "./../async-helpers/tasks";

export default function Dashboard({
  history,
  history: {
    location: { search }
  }
}) {
  const { tasks } = useTaskState();
  const taskDispatch = useTaskDispatch();

  const params = new URLSearchParams(search);
  const page = params.get("page") || 1;
  const perPage = params.get("perPage") || 5;
  const completed = params.get("completed") || undefined;
  const sort = params.get("sort") || undefined;

  const [completedQueryStr, setCompletedQueryStr] = useState("");
  const [completedSortStr, setCompletedSortStr] = useState("");

  useEffect(() => {
    fetchAllTasks(taskDispatch, perPage, page, completed, sort);
  }, [page, perPage, taskDispatch, completed, sort]);

  useEffect(() => {
    if (completed) {
      setCompletedQueryStr(`&completed=${completed}`);
    }
    if (sort) {
      setCompletedSortStr(`&sort=${sort}`);
    }
  }, [completed, sort]);

  return (
    <section>
      <h2>Task List!</h2>
      <FilterBar
        search={search}
        perPage={perPage}
        page={page}
        history={history}
        completed={completed}
        sort={sort}
        completedQueryStr={completedQueryStr}
      />
      <TaskList tasks={tasks} />
      <Pagination
        page={page}
        perPage={perPage}
        completed={completed}
        completedQueryStr={completedQueryStr}
      />
    </section>
  );
}
