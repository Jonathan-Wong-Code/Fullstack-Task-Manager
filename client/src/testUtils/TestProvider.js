import React from "react";
import { UserProvider } from "./../context/user-context";
import { AuthProvider } from "./../context/auth-context";
import { TaskProvider } from "./../context/task-context";

function TestProvider({ children }) {
  return (
    <AuthProvider>
      <UserProvider>
        <TaskProvider>{children}</TaskProvider>
      </UserProvider>
    </AuthProvider>
  );
}

export default TestProvider;
