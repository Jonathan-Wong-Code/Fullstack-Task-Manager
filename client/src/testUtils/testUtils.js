import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import { render } from "@testing-library/react";
import { UserProvider } from "./../context/user-context";
import { AuthProvider } from "./../context/auth-context";
import { TaskProvider } from "./../context/task-context";
import React from "react";
export const renderRouter = (
  ui,
  {
    route = "/",
    history = createMemoryHistory({ initialEntries: [route] }),
    ...options
  }
) => {
  return {
    ...render(<Router history={history}>{ui}</Router>, options),
    history
  };
};

export function renderWithContextRouter(
  ui,
  {
    route = "/",
    history = createMemoryHistory({ initialEntries: [route] }),
    authValue,
    userValue,
    taskValue,
    ...options
  }
) {
  return {
    ...render(
      <AuthProvider value={authValue}>
        <UserProvider value={userValue}>
          <TaskProvider value={taskValue}>
            <Router history={history}>{ui}</Router>
          </TaskProvider>
        </UserProvider>
      </AuthProvider>,
      options
    ),
    history
  };
}
