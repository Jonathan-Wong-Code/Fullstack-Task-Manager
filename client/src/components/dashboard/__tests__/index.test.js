import React from "react";
import { renderWithContextRouter } from "./../../../testUtils/testUtils";
import { cleanup, wait } from "@testing-library/react";
import Dashboard from "..";
import axios from "axios";
afterEach(cleanup);
jest.mock("axios");
const mockData = {
  data: {
    data: {
      tasks: [
        {
          title: "Task 1",
          description: "Task 1",
          completed: false,
          _id: "test-1"
        },
        {
          title: "Task 2",
          description: "Task 2",
          completed: true,
          _id: "test-12"
        }
      ]
    }
  }
};

const mockNumTasks = {
  data: {
    data: {
      complete: 1,
      incomplete: 1
    }
  }
};
describe("<Dashboard />", () => {
  axios.get.mockImplementationOnce(() => Promise.resolve(mockNumTasks));
  axios.get.mockImplementationOnce(() => Promise.resolve(mockData));

  const history = { location: { search: "" } };
  test("it renders", async () => {
    const testUser = { name: "Jon", email: "jon@jon.com" };
    const { container } = renderWithContextRouter(
      <Dashboard history={history} />,
      {
        route: "/dashboard?page=1&perPage=5",
        authValue: testUser
      }
    );

    await wait(() => expect(container).toMatchSnapshot());
  });
});
