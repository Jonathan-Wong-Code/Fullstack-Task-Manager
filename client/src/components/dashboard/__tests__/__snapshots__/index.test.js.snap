// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> it renders 1`] = `
<div>
  <section>
    <h2>
      Task List!
    </h2>
    <form
      action=""
    >
      <label
        for="search"
      >
         Search by title
      </label>
      <input
        id="search"
        name="search"
        type="text"
        value=""
      />
      <label
        for="perPage"
      >
        Number of results
      </label>
      <select
        id="perPage"
        name="perPage"
      >
        <option
          value="5"
        >
          5
        </option>
        <option
          value="10"
        >
          10
        </option>
        <option
          value="15"
        >
          15
        </option>
        <option
          value="20"
        >
          20
        </option>
      </select>
      <label
        for="showCompleted"
      >
        Show tasks:
      </label>
      <select
        id="showCompleted"
        name="showCompleted"
      >
        <option
          value=""
        >
          All
        </option>
        <option
          value="false"
        >
          incomplete
        </option>
        <option
          value="true"
        >
          completed
        </option>
      </select>
      <label
        for="sortBy"
      >
        Sort By:
      </label>
      <select
        id="sortBy"
        name="sortBy"
      >
        <option
          value="-createdAt"
        >
          newest
        </option>
        <option
          value="-completed"
        >
          completed
        </option>
        <option
          value="completed"
        >
          incomplete
        </option>
      </select>
    </form>
    
    <div>
      <ul>
        <li>
          <h3>
            Title: 
            Task 1
          </h3>
          <p>
            Description: 
            Task 1
          </p>
          <label
            for="completed"
          >
            completed?
          </label>
          <input
            id="completed"
            type="checkbox"
            value="false"
          />
          <a
            href="/edit/test-1"
          >
            Edit Task
          </a>
          <button>
            Delete Task
          </button>
        </li>
        <li>
          <h3>
            Title: 
            Task 2
          </h3>
          <p>
            Description: 
            Task 2
          </p>
          <label
            for="completed"
          >
            completed?
          </label>
          <input
            checked=""
            id="completed"
            type="checkbox"
            value="true"
          />
          <a
            href="/edit/test-12"
          >
            Edit Task
          </a>
          <button>
            Delete Task
          </button>
        </li>
      </ul>
    </div>
    <div />
  </section>
</div>
`;
